drivers/ethernet/eth_stm32_hal.c

```diff
@@ -932,3 +1130,285 @@ static struct eth_stm32_hal_dev_data eth0_data = {
+		}
+		heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSSTU;
+		while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSSTU_Msk) {
+			/* spin lock */
+		}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+		ret = 0;
+		irq_unlock(key);
+	}
+
+	return ret;
+}
+
+static int ptp_clock_stm32_rate_adjust(const struct device *dev, float ratio)
+{
+	struct ptp_context *ptp_context = dev->data;
+	struct eth_stm32_hal_dev_data *eth_dev_data = ptp_context->eth_dev_data;
+	ETH_HandleTypeDef *heth = &eth_dev_data->heth;
+	int key, ret;
+	uint32_t addend_val;
+
+	/* No change needed */
+	if (ratio == 1.0) {
+		return 0;
+	}
+
+	key = irq_lock();
+
+	ratio *= eth_dev_data->clk_ratio_adj;
+
+	/* Limit possible ratio */
+	if (ratio < 0.9 || ratio > 1.1) {
+		ret = -EINVAL;
+		goto error;
+	}
+
+	/* Save new ratio */
+	eth_dev_data->clk_ratio_adj = ratio;
+
+	/* Update addend register */
+	addend_val = UINT32_MAX * eth_dev_data->clk_ratio * ratio;
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSAR = addend_val;
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSADDREG;
+	while (heth->Instance->MACTSCR & ETH_MACTSCR_TSADDREG_Msk) {
+		/* spin lock */
+	}
+#else
+	heth->Instance->PTPTSAR = addend_val;
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSARU;
+	while (heth->Instance->PTPTSCR & ETH_PTPTSCR_TSARU_Msk) {
+		/* spin lock */
+	}
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	ret = 0;
+
+error:
+	irq_unlock(key);
+
+	return ret;
+}
+
+static const struct ptp_clock_driver_api api = {
+	.set = ptp_clock_stm32_set,
+	.get = ptp_clock_stm32_get,
+	.adjust = ptp_clock_stm32_adjust,
+	.rate_adjust = ptp_clock_stm32_rate_adjust,
+};
+
+static int ptp_stm32_init(const struct device *port)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(mac));
+	struct eth_stm32_hal_dev_data *eth_dev_data = DEV_DATA(dev);
+	const struct eth_stm32_hal_dev_cfg *eth_cfg = DEV_CFG(dev);
+	struct ptp_context *ptp_context = port->data;
+	ETH_HandleTypeDef *heth = &eth_dev_data->heth;
+	int ret;
+	uint32_t ptp_hclk_rate;
+	uint32_t ss_incr_ns;
+	uint32_t addend_val;
+
+	eth_dev_data->ptp_clock = port;
+	ptp_context->eth_dev_data = eth_dev_data;
+
+	/* Mask the Timestamp Trigger interrupt */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACIER &= ~(ETH_MACIER_TSIE);
+#else
+	heth->Instance->MACIMR &= ~(ETH_MACIMR_TSTIM);
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Enable timestamping */
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	heth->Instance->MACTSCR |= ETH_MACTSCR_TSENA;
+#else
+	heth->Instance->PTPTSCR |= ETH_PTPTSCR_TSE;
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
+	/* Query ethernet clock rate */
+	ret = clock_control_get_rate(eth_dev_data->clock,
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+		(clock_control_subsys_t *)&eth_cfg->pclken,
+#else
+		(clock_control_subsys_t *)&eth_cfg->pclken_ptp,
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+		&ptp_hclk_rate);
+	if (ret) {
+		LOG_ERR("Failed to query ethernet clock");
+		return -EIO;
```

Comment:
Maybe ret itself should be returned here?

Reply:
Returning like this looks fine to me, we clearly know it's an IO error.